### ISSUE TITLE
Support Manual Offset Tracking

### ISSUE THREAD

**Issue by CasperTDK** [rabbitmq/rabbitmq-stream-dotnet-client#3]
Hi. I have made a proof of concept with this SDK and we are only missing:
https://rabbitmq.github.io/rabbitmq-stream-java-client/stable/htmlsingle/#consumer-manual-offset-tracking
to support our use-case for streams.

So any support for this would be wonderful. I am not sure how to contribute, as the java SDK implementation seems quite different. I could look into it, given some pointers however.

**Comment by kjnilsson** [comment]
Hi,

It should just be a matter of implementing a `StoreOffsetRequest` similar to e.g. `TuneRequest` following the spec here: https://github.com/rabbitmq/rabbitmq-server/blob/v3.9.x/deps/rabbitmq_stream/docs/PROTOCOL.adoc#storeoffset

Then implementing a `StoreOffset` method on the `Client` class.

I may get a bit of time next week to do some more work on it.

Cheers

**Comment by kjnilsson** [comment]
NB the StoreOffsetRequest doesn't have a Response so don't wait for one if you decide to have a go at implementing it.

**Comment by CasperTDK** [comment]
@kjnilsson I gave it a shot, but I am stuck around a few key details.
I need to get X messages, handle those, and then commit/store the offset.
It is for batching. I am aware that message offsets have no guarantee to be contiguous

**Comment by acogoluegnes** [comment]
@CasperTDK Make sure to encode the request as [specified in the protocol](https://github.com/rabbitmq/rabbitmq-server/blob/v3.9.x/deps/rabbitmq_stream/docs/PROTOCOL.adoc#storeoffset), here is an [example with the Java client](https://github.com/rabbitmq/rabbitmq-stream-java-client/blob/7754e1f9e355ebdfd9a9d569fd0d91e8531c9d83/src/main/java/com/rabbitmq/stream/impl/Client.java#L1067-L1076).

**Comment by CasperTDK** [comment]
@acogoluegnes I believe I have done so here:
https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/blob/e148cb1ab727bfcf5d7e161f416b16824d4f73ab/RabbitMQ.Stream.Client/StoreOffsetRequest.cs

Or am I misunderstanding something?

**Comment by acogoluegnes** [comment]
If I understand the code correctly, the size for both strings is missing. Check the Java code again:

```java
int length = 2 + 2 + 2 + reference.length() + 2 + stream.length() + 8;
ByteBuf bb = allocate(length + 4);
bb.writeInt(length);
bb.writeShort(encodeRequestCode(COMMAND_STORE_OFFSET));
bb.writeShort(VERSION_1);
bb.writeShort(reference.length()); // size of reference string
bb.writeBytes(reference.getBytes(StandardCharsets.UTF_8)); // reference string
bb.writeShort(stream.length()); // size of stream string
bb.writeBytes(stream.getBytes(StandardCharsets.UTF_8)); // stream string
bb.writeLong(offset);
```

**Comment by kjnilsson** [comment]
WireFormatting.StringSize takes the 2 byte length into account so I think the SizeNeeded is 4 bytes too much atm.

**Comment by CasperTDK** [comment]
Thanks for the hints. I am making progress and QueryOffset request/response now works. 

I am still having issues with `OffsetTypeOffset` in:
`client.Subscribe(stream, new OffsetTypeOffset(queryOffsetResponse.Offset), (ushort)initialCredit,
               new Dictionary<string, string>(), deliverHandler);`

I get a timeout and a very long server error from rabbitmq:
```
2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0> ** State machine <0.19440.0> terminating

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            <<0,0,0,63,0,7,0,1,0,0,0,105,1,0,36,102,101,49,48,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                              98,98,56,45,57,51,100,48,52,101,98,100,52,57,97,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0> ** When server state  = {open,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                           {stream_connection,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            {0,0,0,0,0,65535,44050,5},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            5552,40660,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            1631551285683,<0.19439.0>,#Port<0.247>,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                               {publisher,0,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                <<"3a9340a0-eba6-44c4-89de-2901d1e920bf">>,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                #Ref<0.970995988.3079798785.150228>}},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                               <<"3a9340a0-eba6-44c4-89de-2901d1e920bf">>} =>

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            #{<<"fe10b5c9-7569-4851-abb8-93d04ebd49a0">> =>

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            #{<<"fe10b5c9-7569-4851-abb8-93d04ebd49a0">> =>

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            #Ref<0.970995988.3079798785.150127>,done,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                             [administrator],

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            <<"/">>,opened,0,0,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            #{<<"copyright">> =>

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                              <<"information">> =>

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                              <<"platform">> => <<".NET">>,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                              <<"version">> => <<"0.1.0">>},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                               <<"fe10b5c9-7569-4851-abb8-93d04ebd49a0">>},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                             #Ref<0.970995988.3079667713.150140>},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                           {stream_connection_state,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            false,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                               {consumer,#Port<0.247>,<0.19444.0>,next,0,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                 {cfg,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                  "__fe10b5c9-7569-4851-abb8-93d04ebd49a0_1631551285800003200",

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                  {write_concurrency,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                  undefined,#Fun<osiris_writer.4.69788108>,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                 {read,offset,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                  tcp,user_data},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                 {file_descriptor,prim_file,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                     #Ref<0.970995988.3079798799.46456>,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                    r_buffer =>

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                 undefined},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                                #Ref<0.970995988.3079798785.150223>,#{}}}},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0> ** Reason for termination = error:{case_clause,0}

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0> ** Callback mode = [state_functions,state_enter]

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0> **  [{rabbit_stream_core,parse_request,1,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>      {rabbit_stream_core,'-parse_frames/2-fun-0-',2,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>      {rabbit_stream_core,incoming_data,2,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>      {rabbit_stream_reader,handle_inbound_data,5,

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                             {line,1102}]},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>                            [{file,"src/rabbit_stream_reader.erl"},{line,649}]},

2021-09-13 16:41:31.081564+00:00 [erro] <0.19440.0>      {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>   crasher:

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>     pid: <0.19440.0>

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>     exception error: no case clause matching 0

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>       in call from rabbit_stream_core:'-parse_frames/2-fun-0-'/2 (src/rabbit_stream_core.erl, line 201)

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>       in call from rabbit_stream_reader:handle_inbound_data/5 (src/rabbit_stream_reader.erl, line 1102)

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>       in call from gen_statem:loop_state_callback/11 (gen_statem.erl, line 1194)

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>                   <0.1166.0>,rabbit_stream_sup,<0.1162.0>]

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>     messages: []

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>     dictionary: [{permission_cache,

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>                             <<"fe10b5c9-7569-4851-abb8-93d04ebd49a0">>},

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>                        {{resource,<<"/">>,queue,

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>                         #{},read},

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>                             <<"fe10b5c9-7569-4851-abb8-93d04ebd49a0">>},

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>     trap_exit: true

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>     heap_size: 46422

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0>     reductions: 112709

2021-09-13 16:41:31.085005+00:00 [erro] <0.19440.0> 

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>     errorContext: child_terminated

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>              [{rabbit_stream_core,parse_request,1,

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                                    {line,618}]},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                                   [{file,"src/rabbit_stream_core.erl"},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>               {rabbit_stream_core,incoming_data,2,

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                                    {line,151}]},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                                     [{file,"src/rabbit_stream_reader.erl"},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>               {rabbit_stream_reader,open,3,

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                                      {line,649}]},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                           [{file,"gen_statem.erl"},{line,1194}]},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                         [{file,"proc_lib.erl"},{line,226}]}]}

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                {id,rabbit_stream_reader},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                    {rabbit_stream_reader,start_link,

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                         {acceptor,{0,0,0,0,0,0,0,0},5552},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                           frame_max => 1048576,heartbeat => 60,

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                {restart_type,intrinsic},

2021-09-13 16:41:31.086214+00:00 [erro] <0.19438.0>                {child_type,worker}]

2021-09-13 16:41:31.087051+00:00 [erro] <0.19438.0>     errorContext: shutdown

2021-09-13 16:41:31.087051+00:00 [erro] <0.19438.0>     offender: [{pid,<0.19440.0>},

2021-09-13 16:41:31.087051+00:00 [erro] <0.19438.0>                {mfargs,

2021-09-13 16:41:31.087051+00:00 [erro] <0.19438.0>                        [<0.19439.0>,ranch_tcp,

2021-09-13 16:41:31.087051+00:00 [erro] <0.19438.0>                         #{credits_required_for_unblocking => 12500,

2021-09-13 16:41:31.087051+00:00 [erro] <0.19438.0>                           initial_credits => 50000,transport => tcp}]}},

2021-09-13 16:41:31.087051+00:00 [erro] <0.19438.0>                {shutdown,30000},
```

**Comment by CasperTDK** [comment]
Also, a secondary question. I am understanding the pattern for manual offsets correctly?

e.g. you have a stream, you consume x messages, you store the offset (`StoreOffset`) with the reference.
Later you consume the stream again and use `QueryOffset` + `OffsetTypeOffset` to continue from where you stopped.

**Comment by kjnilsson** [comment]
Yes you first query the offset then you pass the offset returned from the query as the offset spec for the subscription.

I will have a play with your branch tomorrow to see if I can work it out :)

**Comment by CasperTDK** [comment]
@kjnilsson Good news. I became stubborn and found the bug in the OffsetTypeOffset which did not use `span.slice` correctly in `Write`.

So all request/responses now work in my test! 
The last assert is failing, as I am getting 10 messages and not 5 which was expected. 

So either I am misunderstanding how this pattern should work or there is still a bug

**Comment by acogoluegnes** [comment]
The stored offset may be in the middle of a chunk, so when you subscribe you get the whole chunk, so a few extra messages. The client can filter out the first messages, this is what the Java and Go client do, I don't know if it has been implemented yet in the .NET client.

**Comment by CasperTDK** [comment]
@acogoluegnes Ah that makes sense. 
So I guess this:
https://github.com/rabbitmq/rabbitmq-stream-java-client/blob/34eea927c704e95437dd4bc7deaefed9ae19484d/src/main/java/com/rabbitmq/stream/impl/ServerFrameHandler.java#L311

Has to be implemented here:
https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/blob/41f408be4ee9627ea28900b7b4d18d4084866c7f/RabbitMQ.Stream.Client/Deliver.cs#L109

Does that look correct?

**Comment by kjnilsson** [comment]
Hi so the idea is that the current `Client` class is a fairly low level abstraction that doesn't implement any of the "smart" features but instead provide a fairly "raw" API over the protocol.

I havent looked in detail but in your test you have two subscribers incrementing the same mutable value so that may be why you get 10 messages?

**Comment by kjnilsson** [comment]
scrap that I just saw that you reset `messagCount` before the second subscriber. Mutable code can be hard to follow :)

**Comment by kjnilsson** [comment]
but yes so going back to that it could be that all 10 messages that were published ended up inside a single Chunk, so rather than taking the delivery count iterate over the messages and filter out those that are lower than the requested offset.

**Comment by CasperTDK** [comment]
@kjnilsson OK, done!

**Comment by CasperTDK** [comment]
@kjnilsson So now everything is awesome :) ...unless I add a payload to the messages. See #5
